### PR TITLE
System > Advanced > System Tunables

### PR DESCRIPTION
**Pull request by thanegill**

### CODE REVIEW

**Comment by SjonHortensius** [comment]
Thanks for your work! I'll insert a few code-reviews, can you have a look at those?

File: README.md [hunk 1]
```diff
@@ -34,10 +34,13 @@ When finished, don't forget to remove the installation disk from your machine. O
 Disable the dhcp server (on the LAN interface) of your pfSense install and you're good to go. Start the ssh-daemon, login and setup public-key authentication (for the root user). Execute `pkg install rsync` and create a script to upload your changes from your development environment to your pfSense install:
 
 ```bash
-#!/bin/sh
+#! /bin/sh
```
Comment (SjonHortensius) [inline README.md]: Pretty sure that space doesn't belong here

File: usr/local/www/system_advanced_sysctl.php [hunk 2]
```diff
@@ -135,155 +135,93 @@
...
-<?php include("fend.inc"); ?>
-</body>
-</html>
+	</thead>
+	<?php foreach ($tunables as $i => $tunable):
+		if (!isset($tunable['modified']))
+			$i = $tunable['tunable'];
+	?>
+	<tr>
+		<td ondblclick="document.location='system_advanced_sysctl.php?act=edit&amp;id=<?=$i;?>';">
```
Comment (SjonHortensius) [inline usr/local/www/system_advanced_sysctl.php]: The ondblclick attribute should be removed; if you want to keep this functionality, can you move it to a `<script>events.push(function{})</script>` section? It'll automatically be executed onload

File: usr/local/www/system_advanced_sysctl.php [hunk 3]
```diff
@@ -135,155 +135,93 @@
...
+		<td></td>
+		<td></td>
+		<td>
+			<a class="btn btn-xs btn-primary" href="system_advanced_sysctl.php?act=edit">New</a>
+		</td>
+	</tr>
+</table>
+<?php else: ?>
+
+<form action="system_advanced_sysctl.php" method="post" name="iform" id="iform">
```
Comment (SjonHortensius) [inline usr/local/www/system_advanced_sysctl.php]: This form should be converted to php; here's what you could start with:

``` php
require('classes/Form.class.php');
$form = new Form;
$section = new Form_Section('Tunables');

$section->addInput(new Form_Input(
    'Tunable',
    'text',
    $pconfig['tunable']
));

// ...
```

File: README.md (hunk 1 above)
Comment (thanegill) [inline README.md]: Doesn't matter if it's there or not. I personally like it for readability.

File: usr/local/www/system_advanced_sysctl.php (hunk 3 above)
Comment (thanegill) [inline usr/local/www/system_advanced_sysctl.php]: I was going that route at first, but ran into some issues that would either require a logic/interface rewrite or a lot of work to the Form classes. Is there a reason that you're not using a php bootstrap framework? ~~Like this: http://www.phpbootstrap.com/~~ I take that back, the wesite is deceiving compared to the code.

File: usr/local/www/system_advanced_sysctl.php (hunk 2 above)
Comment (thanegill) [inline usr/local/www/system_advanced_sysctl.php]: Yeah, those are not needed.

File: README.md (hunk 1 above)
Comment (SjonHortensius) [inline README.md]: Okay, but if everyone would commit their personal preference this line will be changed a lot ;)

File: usr/local/www/system_advanced_sysctl.php (hunk 3 above)
Comment (SjonHortensius) [inline usr/local/www/system_advanced_sysctl.php]: I had a look at that, but it has no logic for Forms. Also, I think you'll agree that one consistent setup (like we're trying to create) will be easy to migrate to any other framework. I'm pretty happy with the Form setup so far and will definitely help you with any specific issues. What did you run into?

File: README.md (hunk 1 above)
Comment (thanegill) [inline README.md]: Agreed.

File: usr/local/www/system_advanced_sysctl.php (hunk 3 above)
Comment (thanegill) [inline usr/local/www/system_advanced_sysctl.php]: Yes, consistency would be nice. Some of me feels that these efforts are better spent on the python rewrite, as all this PHP will go away as of 3.0, though bootstrap will be nice python or not.

Specifically needed an option for a hidden form field I can hide the field, but not the label. This is as far as I got without needing to modify the Form classes. Also multiple buttons. In this case a save and cancel.

``` php
require('classes/Form.class.php');
$form = new Form;
$section = new Form_Section('Edit Tunable');

$section->addInput(new Form_Input(
    'Tunable',
    'text',
    $pconfig['tunable']
));

$section->addInput(new Form_Input(
    'Description',
    'text',
    $pconfig['descr']
))->forceName('descr');

$section->addInput(new Form_Input(
    'Value',
    'text',
    $pconfig['value']
));

$section->addInput(new Form_Input(
    'id',
    'hidden',
    htmlspecialchars_decode($id)
));

$form->add($section);
print $form;
```

Comment (SjonHortensius) [inline usr/local/www/system_advanced_sysctl.php]: Okay, commit that; and create an issue that you'd like a cancel button. Create an issue for the hidden label as well and possible reference those in your commit. We'll look into it, and apply any modifications from Form into this code.

**Comment by thanegill** [comment]
Updated with suggested changes.
- [ ] Add cancel button to System > Advanced > System Tuneables Edit page #11 
- [ ] Add ability to fully hide an input #10

**Comment by SjonHortensius** [comment]
Looks good. Can you please revert the files besides system_advanced_sysctl.php? I don't want to change files only for making whitespace consistent. If we would, it should be 1 commit that does them all, not 100 small ones :) The Form_Input::forceName fix is already in the repo

If you could rebase all changes in system_advanced_sysctl.php into 1 commit, I'll merge it. Thanks again!

**Comment by thanegill** [comment]
There you go, thanks for the hand holding!

**Pull request merged**